clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -251,13 +252,15 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
             // if we have a response, parse it
             if (response.hasResponse()) {
                 ProduceResponse produceResponse = new ProduceResponse(response.responseBody());
-                for (Map.Entry<TopicPartition, ProduceResponse.PartitionResponse> entry : produceResponse.responses()
-                                                                                                         .entrySet()) {
+                for (Map.Entry<TopicPartition, ProduceResponse.PartitionResponse> entry : produceResponse.responses().entrySet()) {
                     TopicPartition tp = entry.getKey();
                     ProduceResponse.PartitionResponse partResp = entry.getValue();
                     Errors error = Errors.forCode(partResp.errorCode);
                     RecordBatch batch = batches.get(tp);
-                    completeBatch(batch, error, partResp.baseOffset, correlationId, now);
+                    completeBatch(batch, error, partResp.baseOffset, partResp.timestamp, correlationId, now);
+                    if (error == Errors.NONE)
+                        metadata.recordTopicTimestampType(tp.topic(),
```

Comment:
I should have said "If message ever gets re-compressed on broker..." in #2. So we don't over-write timestamp and re-compress just for timestamps.
To answer your question and clarify my suggestions, I am basing my suggestion on the assumption that we don't need to be very exact about timestamp -- meaning +- 1 ms is ok. So my proposal above is assuming that if your scenario happen and the inner message's largest timestamp happened to be the same as the current time on broker and we are using LogAppendTime, this means that producer and broker are in sync, and we don't care whether it is LogAppendTime or CreateTime. One more argument against: what if client sets "bad" timestamp in one of the inner messages. I realized that in your scenario my suggestion does not work well. I propose the following: On broker, when messages get decompressed, if timestamp type == LogApendTime, then set outer message timestamp == current time. Check if max timestamp of inner messages also contains this timestamp, and if so, decrement outer message timestamp by 1 ms. In this case, we know that we don't have the scenario above. On consumer, if outer message timestamp != max of inner messages timestamps, we know that this is LogAppendTime and set all inner message timestamps to outer message timestamp.

**Summary**
I think we either need a cleaner way to get topic's timestamp type on producer or simplify timestamp code by allowing timestamps be not super exact, but no more than +- 1 ms error.
The former I think is better solved by getting timestamp type with topic metadata, but that requires another wire protocol change.
The latter is proposed above, but here is updated version based on the scenario in the previous comment:
1. Producer always sets timestamps (either producer client or KafkaProducer), as if timestamp type == CreateTime. For compressed message: timestamp of the outer message is set to the largest timestamp of the inner messages.
2. Broker will overwrite timestamps if type == LogAppendTime. In case of compressed message: it will overwrite only outer message timestamp, and let inner messages have "wrong" timestamps. If broker happens to overwrite outer message timestamp with with same timestamp, it means that outer timestamp would be equal to max of inner message timestamps. To allow consumer to differentiate between overwritten and not overwritten timestamp, we want to void the case where outer message timestamp == max of inner message timestamps when type == LogAppendTime. So, when overwriting outer message timestamp with same timestamp, we will decrement (or increment) outer message timestamp by 1 ms.
3. When messages get uncompressed on consumer, we know that outer timestamp is either max of inner timestamps (if create time) or outer timestamp was overwritten (and so inner timestamps are not valid anymore). We check if outer (compressed message) timestamp == max of inner message timestamps, and if false, set all inner message timestamps to outer timestamp.